hazelcast/src/main/java/com/hazelcast/jet/impl/processor/AsyncTransformUsingServiceUnorderedP.java

```diff
@@ -316,6 +406,18 @@ public static <C, S, T, K, R> ProcessorSupplier supplier(
     }
 
     private enum Keys {
-        LAST_EMITTED_WM
+        LAST_RECEIVED_WMS
+    }
+
+    /**
+     * Returns `true` if all maps in the given array are empty.
+     */
+    private static <T extends Map<?, ?>> boolean allEmpty(T[] collections) {
```

Comment:
minor: Do we care about cleaning arrays? We increase processor array sizes, but we never shrink them. Maybe we can clear them when allEmpty() returns true.

Reply:
The number of distinct watermark keys is constant and low, in most cases it's 0 or 1. However, the processor doesn't know the keys in advance, so it expands the arrays when it observes a new WM. Normally the arrays will grow to final size very early in the job execution and stay at that size until the end. Shrinking is not needed, as we would grow the arrays very early again after receiving the next WM - WMs typically arrive 10 times per second.